### PR TITLE
Official docker container support.

### PR DESCRIPTION
**Pull request by drecchia**

### CODE REVIEW

**Comment by tananaev** [comment]
I need a bit of explanation of what this all does. Also, next time please send me an email to discuss a new feature before implementing it as we agreed earlier.

**Comment by drecchia** [comment]
Explaining:
- build.sh - Shell script to create the docker container.
- REAME - Some instructions for the final user how to create and run this container.
- Dockerfile - build instructions that will be executed by docker ( not host ) to create the final object.
- start.sh - This file will be copied to the container, its the entry point when the container starts.

File: setup/docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
+FROM java:8-alpine
+
+MAINTAINER Danilo Recchia <danilo.recchia@vortus.solutions>
+
+RUN apk upgrade --update && \
+    apk add --update curl bash && \
```
Comment (tananaev) [inline setup/docker/Dockerfile]: Why do we need both `upgrade` and `add`?

File: setup/docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+ENV JAVA_OPTS -Xms256m -Xmx1024m
+
+COPY ./tmp/traccar.xml /opt/traccar/traccar.xml
+COPY ./tmp/data /opt/traccar/data
+COPY ./tmp/web /opt/traccar/web
+COPY ./tmp/traccar-server.jar /opt/traccar/traccar-server.jar
+COPY ./files/start.sh /opt/traccar/start.sh
+
+RUN chmod +x /opt/traccar/start.sh
```
Comment (tananaev) [inline setup/docker/Dockerfile]: Extra space at the end of the line

File: setup/docker/files/start.sh [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
```
Comment (tananaev) [inline setup/docker/files/start.sh]: Why do we need this file? Traccar creates log file automatically, so we don't need to create it. Basically we need to run java command only.

File: setup/docker/README.txt [hunk 4]
```diff
@@ -0,0 +1,49 @@
+
```
Comment (tananaev) [inline setup/docker/README.txt]: Is this file required? I would prefer to put information on the official website.

File: setup/docker/build.sh [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+export company="tananaev"
+export software="traccar"
+export version=$(head -n 10 ./pom.xml |grep version|cut -d ">" -f2|cut -d"<" -f1)
+
+tmp="./setup/docker/tmp"
+
+mkdir -p ${tmp}/data/database
+cat ./setup/unix/traccar.xml | awk '/web.path/ && !modif { printf("    <entry key=\"web.debug\">true</entry>\n"); modif=1 } {print}' > ${tmp}/traccar.xml
+cp -rf ./database/*.xml ${tmp}/data
+cp -rf ./target/tracker-server-jar-with-dependencies.jar ${tmp}/traccar-server.jar
```
Comment (tananaev) [inline setup/docker/build.sh]: I would prefer to use jar without dependencies and libraries separately to be consistent with other installation packages.

File: setup/docker/build.sh [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+which docker &> /dev/null  || { echo >&2 "Docker package cant be found on path. Aborting."; exit 1; }
+mvn package || { echo >&2 "Maven package has failed. Aborting."; exit 1; }
+
+export company="tananaev"
+export software="traccar"
+export version=$(head -n 10 ./pom.xml |grep version|cut -d ">" -f2|cut -d"<" -f1)
+
+tmp="./setup/docker/tmp"
+
+mkdir -p ${tmp}/data/database
```
Comment (tananaev) [inline setup/docker/build.sh]: Why do we need `database` folder?

File: setup/docker/build.sh [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+cat ./setup/unix/traccar.xml | awk '/web.path/ && !modif { printf("    <entry key=\"web.debug\">true</entry>\n"); modif=1 } {print}' > ${tmp}/traccar.xml
```
Comment (tananaev) [inline setup/docker/build.sh]: What does this command do?

File: setup/docker/build.sh [hunk 8]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+which mvn &> /dev/null || { echo >&2 "Maven package cant be found on path. Aborting."; exit 1; }
+which docker &> /dev/null  || { echo >&2 "Docker package cant be found on path. Aborting."; exit 1; }
+mvn package || { echo >&2 "Maven package has failed. Aborting."; exit 1; }
+
+export company="tananaev"
```
Comment (tananaev) [inline setup/docker/build.sh]: Can we change it to `traccar` or is it somehow linked to github project path?

File: setup/docker/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,23 @@
+FROM java:8-alpine
+
+MAINTAINER Danilo Recchia <danilo.recchia@vortus.solutions>
+
+RUN apk upgrade --update && \
```
Comment (tananaev) [inline setup/docker/Dockerfile]: Can we split it into multiple `RUN` commands?

**Comment by tananaev** [comment]
Finished the review.

File: setup/docker/README.txt (hunk 4 above)
Comment (drecchia) [inline setup/docker/README.txt]: Yes, sure. Create a link on website with content of this docker/README.txt and we and update ./setup/README.txt

File: setup/docker/build.sh (hunk 8 above)
Comment (drecchia) [inline setup/docker/build.sh]: I intend to create I auto build hook in docker hub, triggered by github commit's, so we well need to keep the same as github.

File: setup/docker/build.sh (hunk 7 above)
Comment (drecchia) [inline setup/docker/build.sh]: The command on line 14 add web.debug after web.path, so I dont need to minify js with sencha tools.

File: setup/docker/Dockerfile (hunk 9 above)
Comment (drecchia) [inline setup/docker/Dockerfile]: Multiples RUN create multiples docker layers. Thats not performatic and not recommended.

File: setup/docker/build.sh (hunk 6 above)
Comment (drecchia) [inline setup/docker/build.sh]: Because build is using ./setup/unix/traccar.xml and it points to:
<entry key='database.url'>jdbc:h2:/opt/traccar/data/database</entry>

Comment (drecchia) [inline setup/docker/build.sh]: Since we are using ./setup/unix/traccar.xml, it points to:
<entry key='database.url'>jdbc:h2:/opt/traccar/data/database</entry>

and since we copy ./tmp/data to inside of container, we need to create the db folder before copy all data.

**Comment by tananaev** [comment]
Created a web page:

https://www.traccar.org/docker/

File: setup/docker/build.sh (hunk 6 above)
Comment (tananaev) [inline setup/docker/build.sh]: Last item of the path is the filename, not folder.

**Comment by tananaev** [comment]
Last comment left. Also, please remove the readme file.

I have a small question. Do we really need to declare ports in two places? You declare them in the docker file and then in the run commands as well. Seems a bit weird to me.

File: setup/docker/build.sh (hunk 6 above)
Comment (drecchia) [inline setup/docker/build.sh]: I am not sure if I understand what you mean. Please rephrase.

**Comment by drecchia** [comment]
Following best practices:
https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/

EXPOSE defines ports that can/will be shared with others containers. 
With command line we define which ports will be published (binded) to host interface.

File: setup/docker/build.sh (hunk 6 above)
Comment (tananaev) [inline setup/docker/build.sh]: Database file is stored directly in `/opt/traccar/data` folder. There is never `database` subfolder there.

**Comment by tananaev** [comment]
Publishing ports actually implies expose, but I guess it makes sense to have both.

File: setup/docker/build.sh (hunk 6 above)
Comment (drecchia) [inline setup/docker/build.sh]: Well, then I suggest we split database changelog from database files, since while running external mapped folder we can upgrade container version (with news changelogs) while keeping old database. 
Its ok for you if I point in traccar.xml to this folder ?

Comment (tananaev) [inline setup/docker/build.sh]: We can move changelogs to a separate folder. For example, `schema` or something. But you need to do it everywhere so it's consistent.

Comment (drecchia) [inline setup/docker/build.sh]: Ok, so:
- liquibase files will be moved to ./schema
- h2 files will be places at /opt/traccar/data/
Ok ? 
We will need a little fix at docker instructions, how you wanna do it ?

Comment (tananaev) [inline setup/docker/build.sh]: It would be great to do it as a part of this pull request if you don't mind doing it.

Comment (drecchia) [inline setup/docker/build.sh]: I mean we need to update https://www.traccar.org/docker/ and its not on this project.
Replace this:
docker run -d --name traccar-server -v /my-database-folder:/opt/traccar/data/database -p 5000-5150:5000-5150 -p 8082:8082 tananaev/traccar:[VERSION]

With this:
docker run -d --name traccar-server -v /my-database-folder:/opt/traccar/data -p 5000-5150:5000-5150 -p 8082:8082 tananaev/traccar:[VERSION]

Comment (tananaev) [inline setup/docker/build.sh]: That's not a problem. I will update the website.

**Comment by drecchia** [comment]
Hi, is there anything else missing in this pull request?

**Comment by tananaev** [comment]
Can you please update `setup/package.sh` script to use `schema` folder as well.
